### PR TITLE
ISSUE-2806: Add metrics to dal operations (label with tenant_id and cluster_id) 

### LINKED ISSUE #2806: Add metrics to dal operations (label with  tenant_id and cluster_id)

### PR DESCRIPTION
**Pull request by dantengsky**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

introduce a delegation of DAL  `DALWithMetric`, which intercepts input/output operations and spills simple statistics (bytes read/write) to metrics components

fixing 
- issue #2806

## Changelog

- Improvement
- Not for changelog (changelog entry is not required)

## Related Issues

Fixes #2806

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: query/src/datasources/common/dal_metrics.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        Self {
+            tenant_id: tenant_id.into(),
+            cluster_id: cluster_id.into(),
+            inner: input_stream,
+        }
+    }
+}
+
+pub struct DalWrapper {
+    tenant_id: String,
```
Comment (BohuTANG) [inline query/src/datasources/common/dal_metrics.rs:50]: Seems we can introduce a new struct to `common/metrics/src/recorder.rs` 
```pub struct TenantLabel {
    pub tenant_id:String,
    pub cluster_id:String,
}
```

And here we can use it as:
```
pub struct DalWrapper {
tenant_label: TenantLaber
...
}
```

Comment (dantengsky) [reply query/src/datasources/common/dal_metrics.rs:50]: right!

File: query/src/datasources/common/dal_metrics.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    cluster_id: String,
+    inner: Arc<dyn DataAccessor>,
+}
+
+impl DalWrapper {
```
Comment (BohuTANG) [inline query/src/datasources/common/dal_metrics.rs:55]: `DalWrapper` looks like a strange name and it is rather difficult to understand what it stands for :thinking:

Comment (dantengsky) [reply query/src/datasources/common/dal_metrics.rs:55]: yes, I have the same feeling here. any suggestions?

Comment (dantengsky) [reply query/src/datasources/common/dal_metrics.rs:55]: how about ...  MetricDal... (this is the first name that came up to me, but still "strange" to me)

File: query/src/datasources/common/dal_metrics.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+use common_dal::AsyncSeekableReader;
+use common_dal::DataAccessor;
+use common_dal::InputStream;
+use common_metrics::label_counter_with_val;
+use futures::Stream;
+
+pub const METRIC_DAL_READ_BYTES: &str = "dal.read.bytes";
+pub const METRIC_DAL_WRITE_BYTES: &str = "dal.write.bytes";
+
+struct InputStreamWrapper {
```
Comment (BohuTANG) [inline query/src/datasources/common/dal_metrics.rs:29]: InputStreamWrapper -> InputStreamWithMetric?

Comment (dantengsky) [reply query/src/datasources/common/dal_metrics.rs:29]: :ok_hand:

Comment (dantengsky) [reply query/src/datasources/common/dal_metrics.rs:29]: rename DalWrapper to DalWithMetric as well?

File: query/src/datasources/common/dal_metrics.rs (hunk 2 above)
Comment (BohuTANG) [reply query/src/datasources/common/dal_metrics.rs:55]: MetricDal is ok here :)

File: query/src/datasources/common/dal_metrics.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                    super::dal_metrics::METRIC_DAL_WRITE_BYTES,
+                    stream_len as u64,
+                    self.tenant_id.as_str(),
+                    self.cluster_id.as_str(),
+                )
+            })
+    }
+}
+
+impl futures::AsyncRead for InputStreamWrapper {
```
Comment (BohuTANG) [inline query/src/datasources/common/dal_metrics.rs:110]: Looks like the InputStreamWrapper is a common stream who with metrics.
Would be nice move it to the common/streams as stream_metrics.rs?

Comment (BohuTANG) [reply query/src/datasources/common/dal_metrics.rs:110]: O, the metrics is about DAL.
I think we can make a new file in dal crate called `stream_metrics.rs` out from the dal_metrics file?

Comment (dantengsky) [reply query/src/datasources/common/dal_metrics.rs:110]: :ok:

Comment (dantengsky) [reply query/src/datasources/common/dal_metrics.rs]: > O, the metrics(METRIC_DAL_READ_BYTES...) is about DAL. I think we can make a new file in dal crate called `stream_metrics.rs` out from the dal_metrics file?

yeah, I am trying to move both DalWithMetrics and InputStreamWithMetrics into common/dal

**Comment by dantengsky** [comment]
@BohuTANG  PTAL, `DalWithMetric` and `InputstreamWithMetric` have been moved to `common-dal`, any other suggestions / concerns, pls let me know

**Pull request merged**